blazorApp/Services/VsCodeHelper.cs

```diff
@@ -0,0 +1,130 @@
+    private const string GetLocalResourceRootJsMethod = "getLocalResourceRoot";
+
+    private readonly IJSInProcessRuntime _jsInProcessRuntime;
+    private readonly NavigationManager _navigationManager;
+
+    private NameValueCollection _requiredQueryParameters = default!;
+    private IDisposable _locationChangingRegistration = default!;
+    private string _localResourceRoot = default!;
+
+    private bool _isInitialized;
+
+    public VsCodeHelper(IJSRuntime jsRuntime, NavigationManager navigationManager)
+    {
+        if (jsRuntime is not IJSInProcessRuntime jsInProcessRuntime)
+        {
+            throw new InvalidOperationException(
+                $"The {nameof(VsCodeHelper)} service requires a '{nameof(IJSInProcessRuntime)}' " +
+                $"but got a '{jsRuntime.GetType().Name}' instead.");
+        }
+
+        _jsInProcessRuntime = jsInProcessRuntime;
+        _navigationManager = navigationManager;
+    }
+
+    public void Initialize()
+    {
+        if (_isInitialized)
+        {
+            return;
+        }
+
+        var uriString = _navigationManager.Uri;
+        if (!Uri.TryCreate(uriString, UriKind.RelativeOrAbsolute, out var uri))
+        {
+            throw new InvalidOperationException($"Could not parse the URI '{uriString}'.");
+        }
+
+        if (uri.Query is not { Length: > 0 } queryString)
+        {
+            throw new InvalidOperationException($"The required query string was empty.");
+        }
+
+        _requiredQueryParameters = HttpUtility.ParseQueryString(queryString);
+
+        // VSCode navigates to /index.html by default, which will match the "not found" route
+        // after Blazor starts up. So, we navigate to the home page immediately.
+        // We replace the history entry so you can't navigate back to the "not found" state.
+        _navigationManager.NavigateTo($"/?{_requiredQueryParameters}", new NavigationOptions
+        {
+            ReplaceHistoryEntry = true,
+        });
+
+        // VSCode reads query parameters from the URI to determine how to load resources. Since
+        // navigations would remove that query string, we intercept all navigation attempts,
+        // cancelling them if they don't have the required query parameters. We then re-initiate the navigation
+        // with the required query parameters.
+        _locationChangingRegistration = _navigationManager.RegisterLocationChangingHandler(HandleLocationChangingAsync);
+
+        // Local resources have URIs with an external origin, so relative paths must be converted to have that origin.
+        // We use JS interop to invoke a method that returns the local resource root to help perform this conversion.
+        _localResourceRoot = _jsInProcessRuntime.Invoke<string>(GetLocalResourceRootJsMethod);
+
+        _isInitialized = true;
+    }
+
+    public string AsWebViewUri(string relativeUri)
+    {
+        AssertInitialized();
+        return Path.Combine(_localResourceRoot, relativeUri);
```

Comment:
is this helper giving us a webviewuri? or just something relative to the root (which happens to be a webviewuri?)

Reply:
This is something relative to `baseUrl` in `extension.ts`, which I understand to be the local resource root. So I think it also happens to be what you would get if you were to call the `asWebviewUri()` JavaScript function.